core/src/main/scala/org/bitcoins/core/crypto/TxSigComponent.scala

```diff
@@ -44,6 +46,62 @@ sealed abstract class TxSigComponent {
 
 object TxSigComponent {
 
+  private def inputIndex(inputInfo: InputInfo, tx: Transaction): UInt32 = {
+    tx.inputs.zipWithIndex
+      .find(_._1.previousOutput == inputInfo.outPoint) match {
+      case Some((_, index)) => UInt32(index)
+      case None =>
+        throw new IllegalArgumentException(
+          "Transaction did not contain expected input.")
+    }
+  }
+
+  def apply(
```

Comment:
Why don't we just get away from the carte-blanche `apply()` methods? 

Can't we do something like 

`TxSigComponent.fromWitnessTx()` (i'm guessing this case we can get a more specific type on `InputInfo`)

`TxSigComponent.fromBaseTransaction()` (maybe we can type this with a `RawInputInfo`? 

and then `TxSigComponent.fromBaseTransactionP2SH()` (not sure how this one would look like)

We have some really nasty pattern matching going on (using `_: ScriptPubKey`, which is equivalent to just punting ADTs) which is needed because we aren't narrowing our scope with the parameters to these methods. 

At minimum, (if we don't take my suggestion above) we need to enumerate all of our SPK types instead of using `case _: ScriptPubKey` so we retain pattern match exhaustiveness when we add new scriptpubkey types (say, `WitnessScriptPubKeyV1`)

Reply:
Did this but by type of input info rather than tx. Reason being we could have a witness tx, but the input we are checking is not a segwit input.